pkg/activator/net/revision_backends.go

```diff
@@ -223,45 +223,31 @@ func (rw *revisionWatcher) probePodIPs(ready, notReady sets.String) (succeeded s
 		return rw.healthyPods, true /*no-op*/, false /* notMesh */, nil
 	}
 
-	toProbe := sets.NewString()
-
-	var healthy sets.String
+	healthy := rw.healthyPods.Union(nil)
 	if rw.meshMode != network.MeshCompatibilityModeAuto {
-		// If k8s marked the pod ready before we managed to probe it, and we're
+		// If Kubernetes marked the pod ready before we managed to probe it, and we're
 		// not also using the probe to sniff whether mesh is enabled, we can just
-		// trust k8s and mark it healthy without probing.
-		// TODO: replace with ready.Clone() once we have a recent-enough apimachinery (1.23+).
-		healthy = ready.Union(nil)
-
-		// We still want to probe non-ready pods because we can beat the kubernetes
-		// readiness propagation sometimes.
-		dests = notReady
-	} else {
-		healthy = sets.NewString()
-	}
-
-	for dest := range dests {
-		if rw.healthyPods.Has(dest) {
-			healthy.Insert(dest)
-		} else {
-			toProbe.Insert(dest)
-		}
-	}
-
-	// Short circuit case where the healthy list got effectively smaller.
-	if toProbe.Len() == 0 {
-		return healthy, false, false, nil
+		// trust Kubernetes and mark "Ready" pods healthy without probing.
+		healthy.Insert(ready.List()...)
```

Comment:
`.List()` is ultra expensive as it actually sorts things. `.UnsortedList()` would be slightly better, but `.Union()` will be by far the best, as it'll avoid the intermediary list altogether.

Reply:
union [iterates both lists and inserts in to a new temporary list](https://github.com/knative/serving/blob/f1e419a652cb7447734f86fb0af56e07bbe18b56/vendor/k8s.io/apimachinery/pkg/util/sets/string.go#L114), whereas Insert avoids creating an extra list or iterating the existing healthy set. Probably the most efficient (no extra allocations) though would be:

```suggestion
		for k, _ := range ready { healthy.Insert(k) }
```

wdyt?

Reply:
Indeed so!

Reply:
pushed!